### PR TITLE
Source implementation

### LINKED ISSUE #32: /sources are supported
Desired Behavior
==============================

Stream `/sources` are supported

Benefits
==============================

* Greatly eases adoption
* Immediately unlocks legacy datastores into streaming platform
* Makes deployment / provisioning of source streams very simple

### PR DESCRIPTION
**Pull request by vinayakponangi**
Fixes #32. Sources is a new implementation and should enable registering of CDC sources.

### Added
*  /sources implementation

# PR Checklist Forms

- [x] CHANGELOG.md updated
- [x] Reviewer assigned
- [x] PR assigned (presumably to submitter)
- [x] Labels added (enhancement, bug, documentation)

### CODE REVIEW

File: infra-provider/src/main/java/com/homeaway/streamplatform/streamregistry/provider/InfraManager.java [hunk 1]
```diff
@@ -30,15 +30,15 @@
  *     <li><code>env</code> - Environment, for example staging or production</li>
  *     <li><code>hint</code> - If there are multiple clusters in a vpc, the specific classifier of the cluster. E.g: metrics, logging, etc.</li>
  *     <li><code>type</code> - If there are multiple clusters in a vpc for a type of hint, the specific type of the cluster. E.g. metrics-aggregate or logging-mirror.</li>
- * </ul></p>
```
Comment (cricket007) [inline infra-provider/src/main/java/com/homeaway/streamplatform/streamregistry/provider/InfraManager.java]: End paragraph is still needed.

File: infra-provider/src/main/java/com/homeaway/streamplatform/streamregistry/provider/InfraManager.java [hunk 2]
```diff
@@ -30,15 +30,15 @@
...
+ * </ul>
  * <p>
  * <b><code>ClusterValue</code></b> - Avro encoded value containing containing a <code>Map&lt;String&gt;</code> of Stream Registry cluster properties.
  * This map will be returned to a Stream Registry client at the time of registration.
  * <br>An example of a Confluent Platform Kafka Cluster would contain the following details
  * <ul>
  *     <li><code>bootstrap.servers</code></li>
  *     <li><code>schema.registry.url</code></li>
- * </ul></p>
```
Comment (cricket007) [inline infra-provider/src/main/java/com/homeaway/streamplatform/streamregistry/provider/InfraManager.java]: End paragraph is still needed.

File: docs/docs/architecture/index.md [hunk 3]
```diff
@@ -345,14 +345,14 @@ public interface StreamInfrastructureManager {
    * Validates the configMaps provided for the given cluster are valid.
    */
   void validateCluster(String clusterName,
-                       Map<String, ConfigMap> configMapMap)
+                       Map<String, ConfigMap> configMap)
```
Comment (cricket007) [inline docs/docs/architecture/index.md]: While I agree the name is bad, the paramteter is not just a `ConfigMap` itself.

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/streams/StreamRegistryProducer.java [hunk 4]
```diff
@@ -20,44 +20,34 @@
...
 import org.apache.kafka.clients.producer.ProducerRecord;
 import org.apache.kafka.clients.producer.RecordMetadata;
 
-import com.homeaway.digitalplatform.streamregistry.AvroStream;
-import com.homeaway.digitalplatform.streamregistry.AvroStreamKey;
-import com.homeaway.streamplatform.streamregistry.configuration.TopicsConfig;
-
 @Slf4j
-public class ManagedKafkaProducer implements Managed {
+public class StreamRegistryProducer<K, V> implements StreamProducer<K, V> {
```
Comment (cricket007) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/streams/StreamRegistryProducer.java]: Should this be `K extends AvroStreamKey` for example?

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/streams/GlobalKStreams.java [hunk 5]
```diff
@@ -21,77 +21,72 @@
...
 import org.apache.kafka.streams.state.QueryableStoreTypes;
 import org.apache.kafka.streams.state.ReadOnlyKeyValueStore;
 
-import com.homeaway.digitalplatform.streamregistry.AvroStream;
 import com.homeaway.digitalplatform.streamregistry.AvroStreamKey;
-import com.homeaway.streamplatform.streamregistry.configuration.TopicsConfig;
 
 @Slf4j
-public class ManagedKStreams implements Managed {
+public class GlobalKStreams<T> {
```
Comment (cricket007) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/streams/GlobalKStreams.java]: Should this be `T extends AvroStream`?

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/StreamRegistryApplication.java [hunk 6]
```diff
@@ -124,8 +132,23 @@ public void run(final StreamRegistryConfiguration configuration, final Environme
         kstreamsProperties.put(ROCKSDB_CONFIG_SETTER_CLASS_CONFIG, CustomRocksDBConfig.class);
         TopicsConfig topicsConfig = configuration.getTopicsConfig();
 
-        ManagedKafkaProducer managedProducer = new ManagedKafkaProducer(producerProperties, topicsConfig);
-        ManagedKStreams managedKStreams = new ManagedKStreams(kstreamsProperties, topicsConfig);
+        StreamRegistryProducer<AvroStreamKey, AvroStream> streamProducer = new StreamRegistryProducer(producerProperties, topicsConfig.getProducerTopic());
+        StreamRegistryProducer<AvroStreamKey, Sources> sourceProducer = new StreamRegistryProducer(producerProperties, topicsConfig.getStreamSourceTopic());
```
Comment (cricket007) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/StreamRegistryApplication.java]: Missing diamond? `new StreamRegistryProducer<>(...)`

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/StreamRegistryApplication.java [hunk 7]
```diff
@@ -124,8 +132,23 @@ public void run(final StreamRegistryConfiguration configuration, final Environme
...
+        streamProperties.putAll(kstreamsProperties);
+        streamProperties.put(StreamsConfig.STATE_DIR_CONFIG, "/tmp/streams/kafka-streams");
+
+        GlobalKStreams<AvroStream> streamProcessor = new GlobalKStreams(streamProperties,
+                topicsConfig.getProducerTopic(), topicsConfig.getProducerStateStore(), null);
+
+
+        Properties sourceProperties = new Properties();
+        sourceProperties.putAll(kstreamsProperties);
+        sourceProperties.put(StreamsConfig.STATE_DIR_CONFIG, "/tmp/sources/kafka-streams");
```
Comment (cricket007) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/StreamRegistryApplication.java]: Should the state directories be defined in the config file?

File: config-dev.yaml [hunk 8]
```diff
@@ -38,7 +38,9 @@ kafkaStreamsConfig:
 
 topicsConfig:
   producerTopic: _streamregistry_streams_v0.1
-  stateStoreName: _streamregistry_streams_kv_v0.1
+  producerStateStore: _streamregistry_streams_kv_v0.1
+  streamSourceTopic: _stream_source_command_events_v0.1
+  streamSourceStateStore: _stream_source_command_events_store_v0.1
```
Comment (cricket007) [inline config-dev.yaml]: The config file for the docker image also needs changed

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/impl/KafkaManagerImpl.java [hunk 9]
```diff
@@ -15,11 +15,7 @@
  */
 package com.homeaway.streamplatform.streamregistry.db.dao.impl;
 
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-import java.util.Properties;
+import java.util.*;
```
Comment (cricket007) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/impl/KafkaManagerImpl.java]: No star imports, please

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/impl/SourceDaoImpl.java [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+                Sources updateAvroSources = getAvroSourcesFromJsonList(updatedAvroSource, avroSourcesWithoutTargetItem);
+                kafkaProducer.log(avroStreamKey, updateAvroSources);
+            } else {
+                // add to sources list for an existing stream
+                com.homeaway.digitalplatform.streamregistry.Source updatedAvroSource = getUpdatedAvroSource(givenSource);
+
+                List<com.homeaway.digitalplatform.streamregistry.Source> avroSourcesList = new ArrayList<> ();
+                avroSourcesList.addAll(avroSources
+                        .get()
+                        .getSources());
```
Comment (cricket007) [inline core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/impl/SourceDaoImpl.java]: Why not `avroSourcesList = avroSources.get().getSources()`?

File: infra-provider/src/main/java/com/homeaway/streamplatform/streamregistry/provider/InfraManager.java (hunk 1 above)
Comment (vinayakponangi) [reply infra-provider/src/main/java/com/homeaway/streamplatform/streamregistry/provider/InfraManager.java]: Looks like it's not. The javadoc check was throwing an error. Not exactly HTML like. Here's the official doc - https://www.oracle.com/technetwork/java/javase/documentation/index-137868.html#format.

File: infra-provider/src/main/java/com/homeaway/streamplatform/streamregistry/provider/InfraManager.java (hunk 2 above)
Comment (vinayakponangi) [reply infra-provider/src/main/java/com/homeaway/streamplatform/streamregistry/provider/InfraManager.java]: Same as above.

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/streams/GlobalKStreams.java (hunk 5 above)
Comment (vinayakponangi) [reply core/src/main/java/com/homeaway/streamplatform/streamregistry/streams/GlobalKStreams.java]: Nope. I was making sure this is just doing a basic fetch of X from the store. We might have different stores with a different key later on, but the same logic overall.

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/impl/KafkaManagerImpl.java (hunk 9 above)
Comment (vinayakponangi) [reply core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/impl/KafkaManagerImpl.java]: Yeah. Fixed in my local version. Looks like my checkstyle wasn't set up correctly. Ran into some problems rebasing. I will fix it and push new changes.

File: config-dev.yaml (hunk 8 above)
Comment (vinayakponangi) [reply config-dev.yaml]: Oh. Didn't realize that. Thanks. Will fix.

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/StreamRegistryApplication.java (hunk 7 above)
Comment (vinayakponangi) [reply core/src/main/java/com/homeaway/streamplatform/streamregistry/StreamRegistryApplication.java]: Yeah. Probably a good idea. Will do.

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/StreamRegistryApplication.java (hunk 6 above)
Comment (vinayakponangi) [reply core/src/main/java/com/homeaway/streamplatform/streamregistry/StreamRegistryApplication.java]: Good catch. Fixed.

File: config-dev.yaml (hunk 8 above)
Comment (cricket007) [reply config-dev.yaml]: The Docker config uses environment variables, and a lot of them.

The file here is hard coded and is used by the makefile. 

Not sure a good way to use only one

File: core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/SourceDao.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+package com.homeaway.streamplatform.streamregistry.db.dao;
+
+import java.util.List;
+import java.util.Optional;
+
+import com.homeaway.streamplatform.streamregistry.model.Source;
+
+public interface SourceDao {
+
+    void upsert(Source source);
```
[the remark this answers is not in the document]
Comment (vinayakponangi) [reply core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/SourceDao.java]: Will send some more changes tomorrow.

File: config-dev.yaml (hunk 8 above)
Comment (nathanwalther) [reply config-dev.yaml]: Create a file for dev environment variables, source it prior to `make run` and `make debug`, and use the same `config.yaml`?

**Comment by vinayakponangi** [comment]
Closing this PR after an offline discussion. I will be opening a new one with the new implementation instead.

File: config-dev.yaml (hunk 8 above)
Comment (cricket007) [reply config-dev.yaml]: The one for the docker image needs to be located under `src/main/jib/app`

**Pull request closed without merging**